File: web/src/approval.js
// The approval requests a running turn is waiting on, normalised.
//
// This is the console half of the approval gate, and it is deliberately simpler
// than ask.js, because an approval differs from a question in two ways that both
// cut work:
//
//   - **It is live only.** A question is persisted as a tool call and its result,
//     so a reloaded conversation can render the same card. An approval is not a
//     tool call — it happens *before* one — so a reloaded conversation shows the
//     tool card and its outcome, which is the honest picture: what was asked, and
//     what came of it.
//   - **It has no partial state.** A question can be answered in the person's own
//     words or by choosing; an approval is one of three decisions, and the reason
//     only exists on a refusal.
//
// What it shares with ask.js is the part that matters: the server's outcome event
// is authoritative. A countdown that reaches zero while the server has already
// timed the request out must not leave a card offering buttons.

/** One decision, and what it means on screen. */
export const APPROVAL_ALLOW_ONCE = 'allow_once'
export const APPROVAL_ALLOW_TURN = 'allow_turn'
export const APPROVAL_DENY = 'deny'

/** Card states. Only `pending` accepts input. */
export const APPROVAL_PENDING = 'pending'
export const APPROVAL_SUBMITTING = 'submitting'
export const APPROVAL_SETTLED = 'settled'

function asString(value) {
  return typeof value === 'string' ? value : value === null || value === undefined ? '' : String(value)
}

/** Normalise one line of a request's preview. */
function normalizePreview(raw) {
  const list = Array.isArray(raw) ? raw : []
  const out = []
  for (const entry of list) {
    const text = asString(entry && entry.text).replace(/\r?\n$/, '')
    if (text === '') continue
    const kind = asString(entry && entry.kind)
    out.push({ kind: kind === 'add' || kind === 'del' || kind === 'context' ? kind : 'meta', text })
  }
  return out
}

/**
 * Build the card model from the live `approval` event.
 *
 * Returns null for an event that is not an announcement, so the caller can tell
 * "a new request" from "the outcome of one already on screen" the same way the
 * ask path does.
 */
export function approvalFromEvent(event) {
  const req = event && event.approval
  if (!req || typeof req !== 'object') return null
  const id = asString(req.id)
  if (!id) return null
  return {
    id,
    tool: asString(req.tool),
    capability: asString(req.capability),
    summary: asString(req.summary),
    preview: normalizePreview(req.preview),
    default: asString(req.default),
    // The wait the server is actually applying, so the countdown matches it
    // rather than a number this file invented.
    timeoutMs: Number(req.timeout_ms) > 0 ? Number(req.timeout_ms) : 0,
    status: APPROVAL_PENDING,
    decision: '',
    reason: '',
    source: '',
    error: '',
    announcedAt: Date.now(),
  }
}

/** Apply the server's outcome to a card that is already on screen. */
export function applyApprovalOutcome(card, event) {
  const decision = asString(event && event.approval_decision)
  if (!decision) return card
  card.status = APPROVAL_SETTLED
  card.decision = decision
  card.reason = asString(event && event.approval_reason)
  card.source = asString(event && event.approval_source)
  return card
}

/** True while the card still accepts a decision. */
export function isApprovalOpen(card) {
  return Boolean(card) && (card.status === APPROVAL_PENDING || card.status === APPROVAL_SUBMITTING)
}

/** The requests in a turn that still need an answer. */
export function pendingApprovals(turn) {
  const list = Array.isArray(turn && turn.approvals) ? turn.approvals : []
  return list.filter(isApprovalOpen)
}

/** What the buttons say. */
export function approvalDecisionLabel(decision) {
  switch (decision) {
    case APPROVAL_ALLOW_ONCE:
      return '允许一次'
    case APPROVAL_ALLOW_TURN:
      return '本轮都允许'
    case APPROVAL_DENY:
      return '拒绝'
    default:
      return asString(decision) || '—'
  }
}

/** One line describing how a request ended, for a settled card. */
export function approvalOutcomeLine(card) {
  if (!card || card.status !== APPROVAL_SETTLED) return ''
  const what = approvalDecisionLabel(card.decision)
  if (card.decision !== APPROVAL_DENY) return what
  // A refusal has two very different causes, and they read differently: a person
  // said no, or nobody was there. The second is the one worth spelling out,
  // because it is the gate holding rather than a judgement.
  if (card.source === 'timeout') return '拒绝（等待超时，没有人回应）'
  if (card.source === 'policy') return `拒绝（${card.reason || '按策略处理'}）`
  return card.reason ? `拒绝：${card.reason}` : what
}

/** The hint under the buttons: what each choice means. */
export function approvalHint(card) {
  if (!card) return ''
  if (card.status === APPROVAL_SUBMITTING) return '正在提交…'
  if (!isApprovalOpen(card)) return ''
  return '允许一次只管这一次；本轮都允许在这一次回答结束前不再询问'
}

/**
 * Seconds left on a request, or null when there is no limit to show.
 *
 * It counts down from the wait the server said it is applying, and it never goes
 * below zero: a negative countdown would suggest the request is still alive.
 */
export function approvalSecondsLeft(card, now) {
  if (!card || !isApprovalOpen(card) || !card.timeoutMs) return null
  const elapsed = Math.max(0, (now || Date.now()) - card.announcedAt)
  return Math.max(0, Math.ceil((card.timeoutMs - elapsed) / 1000))
}

File: internal/server/approval.go
package server

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"sync"
	"time"

	"github.com/cloudwego/hertz/pkg/app"
	"go.uber.org/zap"

	"github.com/huan/huan-agent/internal/chat"
	"github.com/huan/huan-agent/internal/tool"
)

// The approval gate's server half.
//
// It is the same shape as the ask_user machinery next door — a hub registered per
// turn, an SSE event that announces the request, an answer that arrives on its own
// request, and a waiter that gives up when the turn ends — with one difference
// that the whole feature depends on:
//
//	**A timeout is a refusal.**
//
// A question that nobody answers leaves the model to decide what to do next,
// which is fine because asking is how it gathers information. An approval that
// nobody answers must not let the action through, because the moment nobody is
// watching is exactly the moment the gate has to hold. Every branch below that
// does not end in an explicit allow ends in a deny.

// DefaultApprovalTimeout bounds how long one write/exec request waits for a
// decision when the deployment does not say otherwise.
//
// It is deliberately shorter than DefaultAskTimeout: a question is the model
// gathering information, while an approval blocks an action, and leaving that
// hanging holds a turn and a half-finished tool call. It is bounded again by the
// turn's own context and by chat.turn_deadline_seconds, which the waiting time
// counts against.
//
// The value matches config.DefaultApprovalTimeout. The two are separate constants
// for the same reason DefaultAskTimeout and DefaultChatAskUserTimeoutSeconds are:
// the server must behave sensibly when built without a config, and the config must
// document a default without importing the server.
const DefaultApprovalTimeout = 5 * time.Minute

// errApprovalGone is what a submission gets when the request it decides is no
// longer waiting: already decided, timed out, cancelled with its turn, or never
// existed. All four are the same 404 for the same reason the question hub gives
// one: a client cannot act on the difference, and telling a stranger which ids
// exist is not worth the precision.
var errApprovalGone = errors.New("approval is no longer pending")

// pendingApproval is one request a running turn is waiting on.
type pendingApproval struct {
	id      string
	session string
	request tool.Request
	// decisions carries the submission to the waiting gate. It is buffered (one
	// slot) so whoever delivers a decision never blocks, even in the moment the
	// waiter has given up.
	decisions chan tool.Decision
	askedAt   time.Time
}

// approvalHub tracks the requests currently waiting for a decision.
//
// It exists for the same reason the question hub does: the decision arrives on a
// different HTTP request from the one the turn is streaming on, so this is the
// only object both can see.
type approvalHub struct {
	mu      sync.Mutex
	pending map[string]*pendingApproval
}

func newApprovalHub() *approvalHub {
	return &approvalHub{pending: map[string]*pendingApproval{}}
}

func (h *approvalHub) register(session string, req tool.Request) *pendingApproval {
	p := &pendingApproval{
		id:        req.ID,
		session:   session,
		request:   req,
		decisions: make(chan tool.Decision, 1),
		askedAt:   time.Now(),
	}
	h.mu.Lock()
	h.pending[req.ID] = p
	h.mu.Unlock()
	return p
}

// peek returns the waiting request without consuming it, so a submission can be
// validated before it decides anything.
func (h *approvalHub) peek(id, session string) (*pendingApproval, bool) {
	h.mu.Lock()
	defer h.mu.Unlock()
	p, ok := h.pending[id]
	if !ok || p.session != session {
		return nil, false
	}
	return p, true
}

// resolve hands a decision to the waiting gate, consuming the request. The
// removal and the hand-off happen under one lock, so of two simultaneous
// submissions exactly one wins and the other is told the request is gone.
func (h *approvalHub) resolve(id, session string, d tool.Decision) error {
	h.mu.Lock()
	p, ok := h.pending[id]
	if !ok || p.session != session {
		h.mu.Unlock()
		return errApprovalGone
	}
	delete(h.pending, id)
	h.mu.Unlock()

	p.decisions <- d
	return nil
}

// forget drops a request its waiter has given up on, so a late decision is
// refused instead of being delivered to nobody.
func (h *approvalHub) forget(id string) {
	h.mu.Lock()
	delete(h.pending, id)
	h.mu.Unlock()
}

// count reports how many requests are waiting. It exists for the tests and for a
// leak check: a hub that grows without bound means a waiter stopped forgetting.
func (h *approvalHub) count() int {
	h.mu.Lock()
	defer h.mu.Unlock()
	return len(h.pending)
}

// turnApprover is the Approver for one turn: it publishes the request on that
// turn's SSE stream and waits for the decision to arrive through the hub.
type turnApprover struct {
	hub     *approvalHub
	session string
	timeout time.Duration
	emit    func(chat.Event)
	logger  *zap.Logger
}

// Approve publishes a request and waits for a decision.
//
// It returns an error only when the machinery failed. "Nobody answered" is a
// DecisionDeny with Source=timeout, not an error, because the gate has to treat
// it as a refusal either way and a person's silence should not be reported to the
// model as a broken tool.
func (a *turnApprover) Approve(ctx context.Context, req tool.Request) (tool.Decision, error) {
	if a.hub == nil {
		// A Server built without New (tests do that) has no registry to park a
		// request in. Failing here with a sentence beats registering into a nil
		// map and taking the process down from inside a tool call.
		return tool.Decision{}, errors.New("server: approval registry is not initialized")
	}

	id, err := newQuestionID()
	if err != nil {
		return tool.Decision{}, fmt.Errorf("分配审批 id: %w", err)
	}
	req.ID = id
	if req.Timeout <= 0 && a.timeout > 0 {
		req.Timeout = a.timeout.Milliseconds()
	}

	pending := a.hub.register(a.session, req)
	// Whichever way this returns, the request stops being decidable. A late
	// submission then gets a 404 rather than flipping a decision already made.
	defer a.hub.forget(id)

	a.emit(chat.Event{Type: chat.EventApproval, Approval: &req})

	timeout := a.timeout
	if req.Timeout > 0 {
		timeout = time.Duration(req.Timeout) * time.Millisecond
	}
	var timerC <-chan time.Time
	if timeout > 0 {
		timer := time.NewTimer(timeout)
		defer timer.Stop()
		timerC = timer.C
	}

	var decision tool.Decision
	select {
	case decision = <-pending.decisions:
		if decision.Source == "" {
			decision.Source = tool.SourceHuman
		}
	case <-timerC:
		decision = tool.Decision{Kind: tool.DecisionDeny, Source: tool.SourceTimeout}
	case <-ctx.Done():
		// The turn is over: the browser went away, 停止 was pressed, or the
		// process is shutting down. Nobody will decide for a turn that no longer
		// exists, and an undecided request must not become an allowed one.
		decision = tool.Decision{Kind: tool.DecisionDeny, Source: tool.SourceTimeout,
			Reason: "本轮已经结束，没有人做出决定"}
	}

	// Tell the client how it ended. A pending card that is never closed would sit
	// there offering buttons for a request nobody is listening to.
	a.emit(chat.Event{
		Type:             chat.EventApproval,
		ApprovalID:       id,
		ApprovalDecision: string(decision.Kind),
		ApprovalReason:   decision.Reason,
		ApprovalSource:   decision.Source,
	})

	if a.logger != nil {
		fields := []zap.Field{
			zap.String("session", a.session),
			zap.String("approval", id),
			zap.String("tool", req.Tool),
			zap.String("capability", string(req.Capability)),
			zap.String("decision", string(decision.Kind)),
			zap.String("source", decision.Source),
			zap.Duration("waited", time.Since(pending.askedAt).Round(time.Millisecond)),
		}
		if decision.Allowed() {
			a.logger.Info("chat: action approved", fields...)
		} else {
			// A refusal is the gate working, not a failure. It is logged at info
			// for a person's decision and at warn for a timeout, because a
			// deployment that sees many timeouts has a wait limit worth raising
			// (or nobody watching, which is what the gate is for).
			if decision.Source == tool.SourceTimeout {
				a.logger.Warn("chat: action refused (no decision)", fields...)
			} else {
				a.logger.Info("chat: action refused", fields...)
			}
		}
	}
	return decision, nil
}

// approvalTimeout is the configured wait limit, or the default.
func (s *Server) approvalTimeout() time.Duration {
	if s.chat.ApprovalTimeout > 0 {
		return s.chat.ApprovalTimeout
	}
	return DefaultApprovalTimeout
}

// handleDecideApproval delivers one card's decision to the turn waiting on it.
//
// Separate from the streaming request for the same reason the question endpoint
// is: that response is already committed to the event stream, and a browser
// cannot write a second body on it. The request id is the join between the two.
func (s *Server) handleDecideApproval(ctx context.Context, c *app.RequestContext) {
	sessionID := c.Param("id")
	approvalID := c.Param("aid")

	var body struct {
		Decision string `json:"decision"`
		Reason   string `json:"reason"`
	}
	if err := c.BindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid JSON body"})
		return
	}
	kind := tool.DecisionKind(strings.TrimSpace(body.Decision))
	if !kind.Valid() {
		c.JSON(http.StatusBadRequest, map[string]string{
			"error": fmt.Sprintf("未知的决定 %q（want allow_once, allow_turn or deny）", body.Decision),
		})
		return
	}
	reason := strings.TrimSpace(body.Reason)
	if len([]rune(reason)) > 500 {
		reason = string([]rune(reason)[:500])
	}

	if s.approvals == nil {
		c.JSON(http.StatusNotFound, map[string]string{"error": errApprovalGone.Error()})
		return
	}
	pending, ok := s.approvals.peek(approvalID, sessionID)
	if !ok {
		c.JSON(http.StatusNotFound, map[string]string{"error": approvalGoneMessage})
		return
	}

	// A reason is only meaningful on a refusal. Dropping it otherwise keeps the
	// audit honest: "rejected because X" and "allowed, and here is X" are
	// different facts, and the model reads the reason as an instruction.
	if kind != tool.DecisionDeny {
		reason = ""
	}

	decision := tool.Decision{Kind: kind, Reason: reason, Source: tool.SourceHuman}
	if err := s.approvals.resolve(approvalID, sessionID, decision); err != nil {
		c.JSON(http.StatusNotFound, map[string]string{"error": approvalGoneMessage})
		return
	}
	if s.logger != nil {
		s.logger.Info("chat: approval recorded",
			zap.String("session", sessionID),
			zap.String("approval", approvalID),
			zap.String("tool", pending.request.Tool),
			zap.String("decision", string(kind)),
			zap.Bool("with_reason", reason != ""),
		)
	}
	c.JSON(http.StatusOK, map[string]bool{"ok": true})
}

// approvalGoneMessage is what a stale submission is told. It names the likely
// causes because the common one is not a bug: the request timed out, and the
// decision the user just clicked arrived after the turn moved on.
const approvalGoneMessage = "这个审批已经不在等待中了（可能已提交、已超时，或本轮已经结束）。超时按拒绝处理，请让模型重新发起。"
